File: js/src/mcp/tools/journeyTools.ts
import { Page } from '@playwright/test';
import { demoUser, TestUser } from '../../lib/fixtures/testUsers.js';
import { journeys, JourneyResult } from '../../lib/journeys/accountJourneys.js';
import { isJourneyId } from '../../lib/journeys/generatedJourneySpecs.js';
import { BrowserController } from './browserTools.js';

export interface JourneyToolParams {
  journeyId: string;
  user?: TestUser;
}

export interface JourneyToolResult extends JourneyResult {}

export interface JourneyTools {
  runJourney(params: JourneyToolParams): Promise<JourneyToolResult>;
}

export interface JourneyBrowserSession {
  getPage(): Promise<Page>;
  dispose(): Promise<void>;
}

export type JourneyBrowserSessionFactory = () => JourneyBrowserSession;

const resolveJourney = (journeyId: string): ((page: Page, user: TestUser) => Promise<JourneyResult>) => {
  if (!isJourneyId(journeyId)) {
    throw new Error(`Unknown journey: ${journeyId}`);
  }
  return journeys[journeyId];
};

export const withIsolatedBrowserSession = async <Result>(
  createSession: JourneyBrowserSessionFactory,
  operation: (session: JourneyBrowserSession) => Promise<Result>
): Promise<Result> => {
  const session = createSession();
  try {
    return await operation(session);
  } finally {
    await session.dispose();
  }
};

export const createJourneyTools = (
  createSession: JourneyBrowserSessionFactory = () => new BrowserController()
): JourneyTools => {
  return {
    runJourney: async ({ journeyId, user }: JourneyToolParams): Promise<JourneyToolResult> => {
      const journey = resolveJourney(journeyId);
      const activeUser = user ?? demoUser;
      return withIsolatedBrowserSession(createSession, async (session) => {
        const page = await session.getPage();
        return journey(page, activeUser);
      });
    }
  };
};

File: py/src/playwright_adventures/mcp_server/tools/journey_tools.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Protocol, TypeVar

from playwright.async_api import Page

from ...journeys import JourneyResult, TestUser, demo_user, is_journey_id, run_journey
from .browser_tools import BrowserSession

Result = TypeVar("Result")


class JourneyBrowserSession(Protocol):
    async def get_page(self) -> Page: ...

    async def close(self) -> None: ...


JourneyBrowserSessionFactory = Callable[[], JourneyBrowserSession]


async def with_isolated_browser_session(
    create_session: JourneyBrowserSessionFactory,
    operation: Callable[[JourneyBrowserSession], Awaitable[Result]],
) -> Result:
    session = create_session()
    try:
        return await operation(session)
    finally:
        await session.close()


class JourneyTools:
    def __init__(self, create_session: JourneyBrowserSessionFactory = BrowserSession) -> None:
        self.create_session = create_session

    async def run_journey(self, journey_id: str, user: TestUser | None = None) -> JourneyResult:
        if not is_journey_id(journey_id):
            raise ValueError(f"Unknown journey: {journey_id}")

        user_obj = TestUser.model_validate(user) if user is not None else demo_user

        async def run(session: JourneyBrowserSession) -> JourneyResult:
            page = await session.get_page()
            return await run_journey(page, journey_id, user_obj)

        return await with_isolated_browser_session(self.create_session, run)
